pallets/oracle/src/lib.rs

```diff
@@ -0,0 +1,511 @@
+		pub fn fetch_price_from_oracle(block_number: T::BlockNumber) {
+			let is_oracle_stopped = <IsOracleStopped<T>>::get();
+			if !is_oracle_stopped {
+				// check that the oracle is NOT stopped
+				let geth_uri = if let Some(geth) = sp_clamor::clamor::get_geth_url() {
+					String::from_utf8(geth).unwrap()
+				} else {
+					log::debug!("Geth URL not set, skipping fetch price from oracle.");
+					return
+				};
+
+				let oracle_address = T::OracleContract::get_contract();
+
+				if let Err(e) = Self::fetch_price(block_number, &oracle_address, &geth_uri) {
+					log::error!("Failed to fetch price from oracle with error: {}", e);
+				}
+
+			} else {
+				log::debug!("The IsOracleStopped flag is set on {:?}. Call stop_oracle(false) to restart it.", is_oracle_stopped);
+				return
+			}
+		}
+
+		fn fetch_price(
+			block_number: T::BlockNumber,
+			contract: &str,
+			geth_uri: &str,
+		) -> Result<(), &'static str> {
+			let req = json!({
+				"jsonrpc": "2.0",
+				"method": "eth_call", // https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_call
+				"params": [
+					{
+					"to": contract,
+						// call `latestRoundData()` function from ChainLink Feed Price contract
+					// `data` is first 4 bytes of `keccak_256(latestRoundData())`, padded - Use https://emn178.github.io/online-tools/keccak_256.html
+					"data": "0xfeaf968c0000000000000000000000000000000000000000000000000000000000000000",
+					},
+					"latest"
+				],
+				 "id": 5, //goerli
+			});
+
+			let req = serde_json::to_string(&req).map_err(|_| "Invalid request")?;
+			log::trace!("Request: {}", req);
+
+			let wait = Timestamp::from_unix_millis(FETCH_TIMEOUT_PERIOD);
+			let response_body = http_json_post(&geth_uri, req.as_bytes(), Some(wait));
+			let response_body = if let Ok(response) = response_body {
+				response
+			} else {
+				return Err("Failed to get response from Ethereum.")
+			};
+
+			let response = String::from_utf8(response_body).map_err(|_| "Invalid response")?;
+			log::trace!("Response: {}", response);
+
+			let v: Value =
+				serde_json::from_str(&response).map_err(|_| "Invalid response - json parse")?;
+			let result = v["result"].as_str().ok_or("Invalid response - no result")?; // Get the latest block number of the Ethereum Blockchain
+			let data = hex::decode(&result[2..]).map_err(|_| "Invalid response - invalid data")?;
+			log::trace!("data: {:?}", data);
+			let data = ethabi::decode(
+				//https://docs.chain.link/docs/data-feeds/price-feeds/api-reference/#latestrounddata
+				&[ParamType::Tuple(vec![
+					ParamType::Uint(80),  // uint80 roundId
+					ParamType::Int(256),  // int256 answer
+					ParamType::Uint(256), // uint256 startedAt
+					ParamType::Uint(256), // uint256 updatedAt
+					ParamType::Uint(80),  // uint80 answeredInRound
+				])],
+				&data,
+			)
+			.map_err(|_| "Invalid response")?;
+
+			let tuple = data[0].clone().into_tuple().ok_or_else(|| "Invalid tuple")?;
```

Comment:
wild cloning, is it necessary? `into` should accept references...

Reply:
The variable `data` returned by `ethabi::decode` contain types `ethabi::Token`, which does not implement the `Copy` trait. 
I also tried to work around it to avoid using `ethabi.into_tuple()` (or `into_int()`) by doing
```rust
 let tuple: Vec<Token> = data[0].try_into().map_err(|_| Error::<T>::InvalidData)?;
```
but 
```rust
the trait `From<ethabi::Token>` is not implemented for `std::vec::Vec<ethabi::Token>`
```

Reply:
hmm fine